python/ray/util/client/server/proxier.py

```diff
@@ -0,0 +1,293 @@
+import time
+from typing import Any, Callable, Dict, Iterator, Optional
+
+import ray.core.generated.ray_client_pb2 as ray_client_pb2
+import ray.core.generated.ray_client_pb2_grpc as ray_client_pb2_grpc
+from ray.util.client.common import (ClientServerHandle,
+                                    CLIENT_SERVER_MAX_THREADS, GRPC_OPTIONS)
+from ray._private.services import ProcessInfo, start_ray_client_server
+
+logger = logging.getLogger(__name__)
+
+CHECK_PROCESS_INTERVAL_S = 30
+
+MIN_SPECIFIC_SERVER_PORT = 23000
+
+MAX_SPECIFIC_SERVER_PORT = 24000
+
+
+def _get_unused_port() -> int:
+    for _ in range(10):
+        port = random.randint(MIN_SPECIFIC_SERVER_PORT,
+                              MAX_SPECIFIC_SERVER_PORT)
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            s.bind(("", port))
+        except OSError:
+            s.close()
+            continue
+        s.close()
+        s.close()
+        return port
+    raise RuntimeError("Unable to succeed in selecting a random port.")
+
+
+def _get_client_id_from_context(context: Any) -> str:
+    metadata = {k: v for k, v in context.invocation_metadata()}
+    client_id = metadata.get("client_id") or ""
+    if client_id == "":
+        logger.error("Client connecting with no client_id")
+        context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+    return client_id
+
+
+@dataclass
+class SpecificServer:
+    port: int
+    process_handle: ProcessInfo
+    channel: "grpc._channel.Channel"
+
+
+class ProxyManager():
+    def __init__(self, redis_address):
+        self.servers: Dict[str, SpecificServer] = dict()
+        self.server_lock = Lock()
+        self.redis_address = redis_address
+
+        self._check_thread = Thread(target=self._check_processes, daemon=True)
+        self._check_thread.start()
+
+    def start_specific_server(self, client_id) -> None:
+        """
+        Start up a RayClient Server for an incoming client to
+        communicate with.
+        """
+        port = _get_unused_port()
+        specific_server = SpecificServer(
+            port=port,
+            process_handle=start_ray_client_server(
+                self.redis_address,
+                port,
+                fate_share=True,
+                server_type="specific-server"),
+            channel=grpc.insecure_channel(
+                f"localhost:{port}", options=GRPC_OPTIONS))
+        with self.server_lock:
+            self.servers[client_id] = specific_server
+
+    def get_channel(self, client_id: str) -> Optional["grpc._channel.Channel"]:
+        """
+        Find the gRPC Channel for the given client_id
+        """
+        client = None
+        with self.server_lock:
+            client = self.servers.get(client_id)
+            if client is None:
+                logger.error(f"Unable to find channel for client: {client_id}")
+                return None
+        grpc.channel_ready_future(client.channel).result(timeout=5)
```

Comment:
I think channel_ready_future will throw an exception.
So there are two kinds of error here: 1) failed to look up; 2) connection failure

For 1) it return None and for 2) it throws an exception.

Could you unify the way we handle errors here?

Reply:
I want to just return None in either context because the caller handles if it is `None`.